File: lc_go/lcGo/dp/lengthOfLIS300.go
package main

import (
	"fmt"
	"sort"
)

func lengthOfLIS(nums []int) int {
	n := len(nums)
	if n == 0 {
		return 0
	}
	//create vector<int> size(n)
	dp := make([]int, n)
	for i := range dp {
		dp[i] = 1
	}
	maxLen := 1
	for i := 1; i < n; i++ {
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] {
				dp[i] = max(dp[i], dp[j]+1)
			}
		}
		maxLen = max(maxLen, dp[i])
	}
	return maxLen
}

func lengthOfLISBinarySearch(nums []int) int {
	n := len(nums)
	if n == 0 {
		return 0
	}
	tails := []int{} // tails[i] 是长度为 i+1 的 LIS 末尾最小值
	for _, num := range nums {
		if len(tails) == 0 || num > tails[len(tails)-1] {
			tails = append(tails, num)
		} else {
			pos := sort.SearchInts(tails, num)
			tails[pos] = num
		}
	}
	return len(tails)
}

func main() {
	nums := []int{10, 9, 2, 5, 3, 7, 101, 18}
	res := lengthOfLISBinarySearch(nums)
	fmt.Println(res)
}

File: lc_cpp/lcCpp/dp/lengthOfLIS300.cpp
//
// Created by 24276 on 2025/3/26.
//

#include <vector>
#include <iostream>

using namespace std;

class length_of_lis300 {
public:
    /**
     * analysis:
     *      1.一维dp 定义状态 dp[i] represented as nums[i] ending subArray(increment)
     *      2.状态转移 define j > i,check i ∈ [0,j - 1]
     * @param nums given nums(Integer)
     * @return find a longest subArray(Increment)
     */
    static int lengthOfLIS(vector<int>& nums) {
        auto n = nums.size();
        if (n == 0) return 0;
        vector<int> dp(n,1);
        int max_len = 1;
        for(int i = 1;i < n;++i) {
            for(int j = 0;j < i;++j) {
                if(nums[j] < nums[i]) {
                    dp[i] = max(dp[i],dp[j] + 1);
                }
            }
            max_len = max(max_len,dp[i]);
        }
        return max_len;
    }

    /**
     * analysis:
     *      1.protect a subarry(tails),tails[i] record as length(i+1) subArray(increment)
     *      希望尽量让子序列增长，但如果当前数字比所有末尾都大，就加到后面；如果比某些末尾小，就替换掉最接近的大值，
     * @param nums given nums
     * @return
     */
    static int LengthOfLISBINARYSEARCH(vector<int>& nums) {
        auto n = nums.size();
        if (n == 0) return 0;
        vector<int> tails;
        for(auto num : nums) {
            if(tails.empty() || num > tails.back()) {
                tails.push_back(num);
            }else {
                //找到第一个 >= num的subscript下标。替换为当前
                int pos = lower_bound(tails.begin(), tails.end(), num) - tails.begin();
                tails[pos] = num;
            }
        }
        return tails.size();
    }
};

int main() {
    length_of_lis300 solution;
    vector<int> nums= {10,9,2,5,3,7,101,18};
    cout << solution.LengthOfLISBINARYSEARCH(nums) << endl;
    return 0;
}

